Redirect Wildcard Doc (#2026)
FusionAuth/fusionauth-site: site/docs/v1/tech/oauth/url-validation.adoc

---
layout: doc
title: URL Validation

> description: A description OAuth URL validation policies in FusionAuth

Reviewer: ```suggestion
description: OAuth URL validation policies in FusionAuth
```

navcategory: login-methods
---

[NOTE.since]
====
Available since 1.43.0
====

== Overview

Beginning in version `1.43.0` FusionAuth provides support for wildcards in OAuth 2.0 redirect URLs and origin URLs. This document provides details on where wildcards are allowed in configured values and the valid replacement patterns for wildcards in each position.

* <<URL Validation Policy>>
* <<Allowed Wildcard Positions>>
** <<Domain>>
** <<Port>>
** <<Path Segments>>
** <<Query String Values>>
* <<Wildcard Replacement Patterns>>
** <<Domain>>
** <<Port>>
** <<Path Segments>>
** <<Query String Values>>

== URL Validation Policy

> In order to configured authorized redirect and origin URLs with wildcards, the [field]#URL validation# setting must be configured to `Allow wildcards` under [breadcrumb]#Applications -> Edit Application -> OAuth#.

Reviewer 2: > In order to configure

> See the link:/docs/v1/tech/apis/applications[Application API] or [Application OAuth Configuration](/docs/v1/tech/core-concepts/applications#oauth) for details.

Reviewer 2: This second link is broken.

Author: I didn't think `link:` worked when you wanted to link to a specific section on the page. That resolved it, but what was wrong with the syntax I had? I've seen it work in other places.

Reviewer: I think he is talking about `[Application OAuth Configuration](/docs/v1/tech/core-concepts/applications#oauth)`

Author: Yeah. I changed it to a `link:` this morning. I thought the `[text](url)` pattern was another way to create links in AD/MD.

== Allowed Wildcard Positions

> In order to maintain security while allowing the flexibility of wildcards, FusionAuth limits the position and number of wildcards that are allowed in the configured authorized redirect and origin URLs. The `*` is used to denote a wildcard.

Reviewer: ```suggestion
In order to maintain security while allowing the flexibility of wildcards, FusionAuth limits the position and number of wildcards that are allowed in the configured authorized redirect and origin URLs. The asterisk character, `*`, is the wildcard character.
```

=== Domain

The domain of a configured URL allows 0 or 1 wildcards in the domain portion of the URL. If the domain contains a wildcard, it must meet all of the following requirements:

>  * The domain must contain at least three segments

Reviewer: ```suggestion
 * The domain must contain at least three segments.
```

>  * The wildcard may only appear in the host portion of the domain (left-most subdomain)

Reviewer: ```suggestion
 * The wildcard may only appear in the host portion of the domain (left-most subdomain).
```

Reviewer: Is it host or hostname? Looks like it is `host`. https://www.rfc-editor.org/rfc/rfc3986#section-3.2.2

>  * The wildcard can be a full or partial replacement of the host

Reviewer: ```suggestion
 * The wildcard can be a full or partial replacement of the host.
```

The following table provides some examples of valid and invalid wildcard patterns.

[cols="45m,10,45"]
.Domain Wildcards

> |===

Reviewer 2: We have a similar table here https://site-local.fusionauth.io/docs/v1/tech/guides/breached-password-detection#match-mode in `Table 1. What Is Allowed In Each Match Mode`, we should be able to copy the way we did check marks and x's.

| Example | Valid | Reason

> |\https://*.example.com

Reviewer: What about `https://*.98.1.1`? Does that match any numbers?

Author: A wildcard is not allowed when the host is an IP address. I can add an example.

|✓
|&nbsp;

|\https://blah*.example.com
|✓
|&nbsp;

|\https://*.com
|✕
|The domain only contains two segments

|\https://auth.*.com
|✕
|The wildcard does not appear in the host

|\https://pass:[*]midpass:[*].example.com
|✕
|The domain contains multiple wildcards

|===

=== Port

The port number can be specified as a wildcard. There is no partial wildcard support for the port number.

[cols="45m,10,45"]
.Port Wildcards
|===
| Example | Valid | Reason

|\https://example.com:*
|✓
|&nbsp;

|\https://example.com:4*
|✕
|Partial wildcards are not allowed for the port number

|===

=== Path Segments

Wildcards are allowed in path segments with the following restrictions:

>  * There can be no more than one wildcard _per path segment_

Reviewer: ```suggestion
 * There can be no more than one wildcard _per path segment_.
```

>  * The wildcard can be a full or partial replacement of the path segment

Reviewer: ```suggestion
 * The wildcard can be a full or partial replacement of the path segment.
```

The following table provides some examples of valid and invalid wildcard patterns in the path.

[cols="45m,10,45"]
.Path Wildcards
|===
| Example | Valid | Reason

|\https://example.com/path/*/resource
|✓
|&nbsp;

|\https://example.com/p*/to/resource
|✓
|&nbsp;

|\https://example.com/pass:[*]/par*tial/pass:[*]
|✓
|&nbsp;

|\https://example.com/path/pass:[*]midpass:[*]/resource
|✕
|The path segment contains multiple wildcards

|===

=== Query String Values

Wildcards are allowed in query string values with the following restrictions:

>  * Partial wildcards are not allowed for query string values

Reviewer: ```suggestion
 * Partial wildcards are not allowed for query string values.
```

Wildcards are _not_ allowed in query string names. The following table provides some examples of valid and invalid wildcard patterns in the query string.

[cols="45m,10,45"]
.Query String Wildcards
|===
| Example | Valid | Reason

|\https://example.com?foo=*
|✓
|&nbsp;

|\https://example.com?foo=pass:[*]&bar=pass:[*]&baz=blah
|✓
|&nbsp;

|\https://example.com?foo=par*tial
|✕
|Partial wildcard replacement is not allowed for query string values

|\https://example.com?*=blah
|✕
|Wildcards are not allowed in query string names

|===

== Wildcard Replacement Patterns

> The position where wildcards are allowed in configured values is just one half of the puzzle. Wildcards in each portion of the URL have different rules for the replacement values. Please note that allowed replacement values may not produce a valid URL. This section provides details on the allowed replacements for wildcards in each portion of the URL. A wildcard in the configured value must be replaced with at least one character in the concrete value.

Reviewer: What is a concrete value? This is the first I've heard of that term.

Author: I thought that was common parlance. A configured authorized URL with a wildcard is a pattern or template. The actual value coming in on `redirect_uri` or the `Origin` header would be the concrete value that is matched against the pattern.

I'm open to suggestions for changing this verbiage.

Reviewer 2: could we say:

> .. in the value.

Or 

> ... in the literal value.

Reviewer: Was confusing to me :) 

Maybe: 

`A wildcard in the configured value must be replaced with at least one character in the concrete value.` => 

`Each wildcard in the configured value must match one or more characters. Matches against empty strings will fail.`

=== Domain

> Replacements for wildcards in the domain portion of the URL must not contain `.`, `:`, `/`, or `?` characters. The following table lists examples of valid and invalid replacements for valid wildcard patterns.

Reviewer: Do we test for url escaped values? `%2F`

[cols="25m,25m,10,40"]
.Domain wildcard replacement
|===
| Pattern | Value | Valid | Reason

|\https://*.example.com
|\https://login.example.com
|✓
|&nbsp;

|\https://auth*.example.com
|\https://auth2.example.com
|✓
|&nbsp;

|\https://auth*.example.com
|\https://auth.example.com
|✕
|The value does not contain a character to replace the `*`

|\https://*.example.com
|\https://auth.customer.example.com
|✕
|The replacement contains a `.` character

|===

=== Port

> Replacement values for wildcards in the port portion of the URL must consist of one or more digits.

Reviewer: ```suggestion
Replacement values for wildcards in the port portion of the URL must consist of one or more decimal digits.
```

[cols="25m,25m,10,40"]
.Port wildcard replacement
|===
| Pattern | Value | Valid | Reason

|\https://example.com:*
|\https://example.com:2012
|✓
|&nbsp;

|\https://example.com:*
|\https://example.com:80b
|✕
|The replacement value contains a non-numeric character

|===

=== Path Segments

> Replacement values for wildcards in a path segment of the URL must not contain `/` or `?` characters.

Reviewer: Do we test for URL escaped values? `%2F`?

Reviewer: As I'm reading the docs, I was wondering if `https://example.com/path/*/resource` would match `https://example.com/path%2fto%2Fthe/resource` . I just tested on sandbox and it fails. But would probably be a good test to have.

Author: I think that fails because it doesn't have `/path/`. Pretty sure that `/path/to%2Fthe/resource` would pass.

I have a task to add some tests for the following:
* All of the examples in doc (valid and invalid)
* The [example](https://inversoft.slack.com/archives/C0407G6V5/p1678138231518359) you called out in Slack
* A couple of these escape sequences

[cols="25m,25m,10,40"]
.Path segment wildcard replacement
|===
| Pattern | Value | Valid | Reason

|\https://example.com/path/*/resource
|\https://example.com/path/to/resource
|✓
|&nbsp;

|\https://example.com/p*/to/resource
|\https://example.com/path/to/resource
|✓
|&nbsp;

|\https://example.com/pass:[*]/par*tial/pass:[*]
|\https://example.com/path/partotial/resource
|✓
|&nbsp;

|\https://example.com/path/*/resource
|\https://example.com/path/to/the/resource
|✕
|The replacement value contains a `/`

|\https://example.com/path/*
|\https://example.com/path/resource?foo=bar
|✕
|The replacement value contains a `?`

|\https://example.com/pass:[*]/par*tial/pass:[*]
|\https://example.com/path/partial/resource
|✕
|The segment `partial` does not contain a replacement character for the wildcard

|===

=== Query String Values

> Replacement values for query string values must not contain the `&` character.

Reviewer: Do we test for a url escaped `&`?

[cols="25m,25m,10,40"]
.Query string value wildcard replacement
|===
| Pattern | Value | Valid | Reason

|\https://example.com?foo=*

> |\https://example.com?foo=bar

Reviewer: Does this match `https://example.com?bar=baz&foo=bar`?

I'd guess no, since `bar=baz` is not matched.

Author: That's correct. Added an example.
